implement fw-based Phase2 EG standalone barrel

skkwan opened the pull request:
#### PR description:

This PR implements the firmware-based Phase 2 EGamma standalone barrel emulator. The output collections are of the same data type as the TDR (previous version) non-firmware-based Phase 2 emulator, namely l1t::EGamma, l1tp2::CaloCrystalClusters, and 1tp2::CaloTower. 

Some firmware-style code is refacored into abstracted functions and classes, compared to the firmware. 

This PR also corresponds to [PR#1069 to cms-l1t-offline](https://github.com/cms-l1t-offline/cmssw/pull/1069). 

#### PR validation:

Efficiencies and rates for the EG clusters produced in the firmware-based emulator are [compared](https://github.com/cms-l1t-offline/cmssw/pull/1069#issuecomment-1479547259) with the previous TDR emulator. 

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

N/A

skkwan committed:
implement fw-based Phase2 EG standalone barrel

Produce l1t::EGamma, l1tp2::CaloCrystalClusters, and 1tp2::CaloTower
output collections in same style as TDR (previous) emulator.

Address code formatting comments.
Refactor firmware-style code to abstracted functions and classes.

aloeliger commented:
@skkwan Thanks! Congratulations on getting this ready! Any commits that are made in central CMSSW review we would like cherry-picked back to the branch with the PR open to the l1-fork. Hopefully there shouldn't be much after the review.

For anyone please, please also take a look at the review and suggestions made in the [l1-offline-fork](https://github.com/cms-l1t-offline/cmssw/pull/1069)

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/34940

ERROR: Build errors found during clang-tidy run.
```
L1Trigger/L1CaloTrigger/interface/Phase2L1CaloEGammaEmulator.h:347:18: error: array index 5 is past the end of the array (which contains 5 elements) [clang-diagnostic-array-bounds]
    linkECAL() { crystalE[CRYSTALS_IN_TOWER_ETA][CRYSTALS_IN_TOWER_PHI] = {}; }
                 ^
L1Trigger/L1CaloTrigger/interface/Phase2L1CaloEGammaEmulator.h:343:5: note: array 'crystalE' declared here
--
L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloEGammaEmulator.cc:360:28: error: variable-sized object may not be initialized [clang-diagnostic-error]
    int towerEtaBoundaries[nRegionBoundariesEta][2] = {{15, 14}, {12, 11}, {9, 8}, {6, 5}, {3, 2}};
                           ^
Suppressed 1432 warnings (1431 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

skkwan committed:
fix: build-bot comments

Fix build-bot comments on constructor/initializers

skkwan committed:
fix: remove unnecessary cmsRun config files from PR

Remove cmsRun config files which are not unit tests from the PR.

skkwan committed:
fix: fix scram b code-checks

skkwan committed:
fix: fix scram b code-format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/34954

- This PR adds an extra 60KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

aloeliger commented:
please abort

skkwan committed:
fix: add hls to BuildFile

Add hls to BuildFile since plugins and interface depend on ap_int.h

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/34955

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

cmsbuild commented:
Pull request #41224 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

aloeliger commented:
@skkwan Could you please double check that all includes needed by your header have had appropriate includes added to the build files?

cmsbuild commented:
-1

**Failed Tests**:   HeaderConsistency 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/31693/summary.html
**COMMIT**: 5c50de35b6bcec8b55656a2392327fc65b5c41a7
**CMSSW**: CMSSW_13_1_X_2023-03-30-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41224/31693/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 8 lines from the logs
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3554236
 - DQMHistoTests: Total failures: 113
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3554101
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-30-1100+81ce5a/56341/triggerResults): found differences in 3 / 47 workflows

aloeliger commented:
Seeing this is my first time learning about `scram b check-headers`. I'll add something for it into the L1 offline fork PR testing scripts I am developing for the future so we don't get caught in this sort of thing in the future.

@skkwan you say this builds alright when you compile it from your side?

I don't see any documentation in the scram twiki or under the PR testing protocol for the header checks. I know the documentation in this collaboration is permanently out of date (a fact I complain about but don't help much...) but is this error common enough to introduce it might be good to mention it in the _nominal_ PR testing procedure?

skkwan committed:
fix: fix incomplete header and build files

Fixes errors that were thrown by scram build -k -j 16 check-headers

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/34977

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

cmsbuild commented:
Pull request #41224 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/31709/summary.html
**COMMIT**: a16713639736d8a3e492423347ac288c99448f75
**CMSSW**: CMSSW_13_1_X_2023-03-30-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41224/31709/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 1 lines from the logs
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3553709
 - DQMHistoTests: Total failures: 112
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3553575
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-30-2300+81ce5a/56356/triggerResults): found differences in 3 / 47 workflows

aloeliger commented:
@epalencia Anything for stephanie? Otherwise we've had this available and under review for a little while now and we can pass review to upgrade and orp.

skkwan commented:
Hello @upgrade-l2 @srimanob , are there any comments from the side of upgrade-l2 for this PR? Thank you! I tried to tag upgrade-l2 but not sure if it worked.

aloeliger commented:
@cms-sw/upgrade-l2 Are there any comments for @skkwan?

AdrianoDee commented:
@aloeliger given this is a 4k lines PR it could take a bit more than a couple of days to review it. Do you need it urgently for some reason?

aloeliger commented:
@AdrianoDee L1 has phase-2 deadlines essentially now. It's too late for those at this point, but I think the phase 2 developers will have follow-up deadlines at the end of the month. In particular, upstream emulators have been stressing to me the necessity of having this particular emulator integrated.

skkwan committed:
add digitized collections, squash all changes not in cmssw pr

Add digitized collections and squash all changes not in
https://github.com/cms-sw/cmssw/pull/41224.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/35152

- This PR adds an extra 52KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #41279, #40563

cmsbuild commented:
Pull request #41224 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/31950/summary.html
**COMMIT**: 1e5901adc1ab67c51008f821d89d60c5814a3be9
**CMSSW**: CMSSW_13_1_X_2023-04-12-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41224/31950/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 8 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3459609
 - DQMHistoTests: Total failures: 77
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3459510
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+81ce5a/56502/triggerResults): no differences found

aloeliger commented:
Recent commit by @skkwan adds "bit accurate"/"digitized" collections agreed upon with upstream emulators alongside un-digitized collection provided in prototype (to be removed after suitable period of validation).

skkwan committed:
squash commits fixing digitized collections

Squash commits not in https://github.com/cms-sw/cmssw/pull/41224 addressing:
- use two's-complements in DigitizedClusterGT
- fix LSB value in DigitizedClusterGT

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/35227

- This PR adds an extra 80KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #41279, #40563

cmsbuild commented:
Pull request #41224 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/32077/summary.html
**COMMIT**: e0980db87610d3edfa0b95faece78a11559226e9
**CMSSW**: CMSSW_13_1_X_2023-04-20-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41224/32077/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 2 lines to the logs
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460915
 - DQMHistoTests: Total failures: 76
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460817
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-20-1100+81ce5a/56610/triggerResults): found differences in 1 / 46 workflows

epalencia commented:
> @epalencia Do you have any comments for Stephanie after adding the digitized collections, or is it time that L1 turned over review to other groups who haven't developed tunnel vision on this PR?

No, it looks good. I'll sign now.

AdrianoDee commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/32759/summary.html
**COMMIT**: e0980db87610d3edfa0b95faece78a11559226e9
**CMSSW**: CMSSW_13_2_X_2023-05-23-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41224/32759/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 4 lines from the logs
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3221591
 - DQMHistoTests: Total failures: 75
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3221494
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-23-1100+81ce5a/57171/triggerResults): found differences in 3 / 46 workflows

AdrianoDee commented on L1Trigger/L1CaloTrigger/interface/Phase2L1CaloEGammaEmulator.h:
```diff
@@ -0,0 +1,1738 @@
+
+  class etaStrip_t {
+  public:
+    ecaltp_t cr0;
```
These `crX` here, the `etaStripX` (in `ecalRegion_t`) and the `pkX` may be transferred to an array-like structure? I'm asking because in the `p2eg::initStructure`, `p2eg::getPeakBin20N`, `p2eg::getPeakBin15N`, `p2eg::getClusterPosition` and `p2eg::getECALTowersEt` are quite cumbersome because of this choice (seems to me they could be simplified with a loop).

skkwan replied:
Hi @AdrianoDee, these parts of the code came from the firmware; our plan is to make minimal changes to the firmware blocks to preserve the ease of translating firmware back and forth.

AdrianoDee replied:
Fine by me, thanks for the clarification.

AdrianoDee reviewed (commented):
Done with the review.

skkwan committed:
add fillDescriptions and rename header

Add fillDescription to Phase2L1CaloEGammaEmulator.cc, rename header
file to be more descriptive, and move #includes to after #ifndef guard.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/35654

- This PR adds an extra 52KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/interface/Phase2L1CaloEGammaEmulator.h:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Modified: 28071b1ab59bd5ca3cdc1c3cce3af742ae309c39, abe504273ad97be1942dbe659a2cdf1c8a21060c, a343481398ea69b17017dae68956e928435004a4, a16713639736d8a3e492423347ac288c99448f75, 1e5901adc1ab67c51008f821d89d60c5814a3be9
    - Deleted: b6eb07bd58c80ba8a464bce83df49c113f41db1c
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/35654/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/35654/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

aloeliger commented:
@skkwan Any progress with this PR?

aloeliger committed:
Update L1EGammaCrystalsEmulatorProducer.cc format

fix format issue

skkwan committed:
Merge pull request #1 from aloeliger/patch-1

Update L1EGammaCrystalsEmulatorProducer.cc format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41224/35851

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - L1Trigger/L1CaloTrigger/interface/Phase2L1CaloEGammaEmulator.h:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Modified: 28071b1ab59bd5ca3cdc1c3cce3af742ae309c39, abe504273ad97be1942dbe659a2cdf1c8a21060c, a343481398ea69b17017dae68956e928435004a4, a16713639736d8a3e492423347ac288c99448f75, 1e5901adc1ab67c51008f821d89d60c5814a3be9
    - Deleted: b6eb07bd58c80ba8a464bce83df49c113f41db1c
  - L1Trigger/L1CaloTrigger/test/test_Phase2L1CaloEGammaEmulator.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b
  - L1Trigger/L1CaloTrigger/test/test_L1EGammaCrystalsProducer.py:
    - Added: 0caac299ef54b90abb600f30533b99cd4cbf8f13
    - Deleted: 2ef4adf5e0e64f5b225e3daf024d3577632c394b

cmsbuild commented:
Pull request #41224 was updated. @epalencia, @AdrianoDee, @srimanob, @cmsbuild, @aloeliger can you please check and sign again.

aloeliger commented:
+l1

AdrianoDee commented:
please test
(being super pedantic here, as soon as the tests are good I'll sign)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/33159/summary.html
**COMMIT**: f9e26714e1e4f46052edf0cfaa0a466df6aff347
**CMSSW**: CMSSW_13_2_X_2023-06-14-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41224/33159/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 28 lines from the logs
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3196062
 - DQMHistoTests: Total failures: 78
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3195962
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-14-1100+81ce5a/57490/triggerResults): found differences in 3 / 46 workflows

AdrianoDee commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented on L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloEGammaEmulator.cc:
```diff
@@ -0,0 +1,753 @@
+                                  });
+    desc.add<edm::ParameterSetDescription>("calib", psd0);
+  }
+  descriptions.add("l1tPhase2L1CaloEGammaEmulator", desc);
+  // or use the following to generate the label from the module's C++ type
+  //descriptions.addWithDefaultLabel(desc);
```
As such, you end up with two configuration files named l1tPhase2L1CaloEGammaEmulator_cfi.py, and both accessible in python as L1Trigger.L1CaloTrigger.l1tPhase2L1CaloEGammaEmulator_cfi. This generates confusion (especially if at some time any of them differs from the other) and must be avoided.

If these configurations are not expected to change, you can remove the explicit file L1Trigger/L1CaloTrigger/python/l1tPhase2L1CaloEGammaEmulator_cfi.py

Since I imagine that is not the case, then:
```suggestion
  descriptions.addWithDefaultLabel(desc);
```

After that, you can use this configuration description only to validate the one in L1Trigger/L1CaloTrigger/python/l1tPhase2L1CaloEGammaEmulator_cfi.py

Or, you can replace the l1tPhase2L1CaloEGammaEmulator configuration in it with
```PYTHON
from L1Trigger.L1CaloTrigger.phase2L1CaloEGammaEmulator_cfi import phase2L1CaloEGammaEmulator
l1tPhase2L1CaloEGammaEmulator = phase2L1CaloEGammaEmulator.clone()
```
where possible modified parameters can be written as arguments of the clone()

aloeliger replied:
@perrotta I have @skkwan sitting in my office, and we want to clarify a point of preference here.

To my understanding right now the `_cfi` does/did two things. First it configured (redundantly before the change) the EGamma emulator, but second, we have this era modification hanging around:

```python
from Configuration.ProcessModifiers.premix_stage2_cff import premix_stage2
premix_stage2.toModify(l1tPhase2L1CaloEGammaEmulator,
    ecalTPEB = "DMEcalEBTriggerPrimitiveDigis",
    hcalTP = "DMHcalTriggerPrimitiveDigis",
)
```

After looking at this for a minute, I think there are two possible ways this end up:

```python
import FWCore.ParameterSet.Config as cms

from L1Trigger.L1CaloTrigger.phase2L1CaloEGammaEmulator_cfi import phase2L1CaloEGammaEmulator
l1tPhase2L1CaloEGammaEmulator = phase2L1CaloEGammaEmulator.clone(    
    ecalTPEB = "DMEcalEBTriggerPrimitiveDigis",
    hcalTP = "DMHcalTriggerPrimitiveDigis",
)
```

And then we expect that whoever uses the phase2_premix era uses this module instead of the auto-generated `phase2L1CaloEGammaEmulator`. This seems a bit prone to failure. Alternatively, I guess one could do this:

```python
import FWCore.ParameterSet.Config as cms

from L1Trigger.L1CaloTrigger.phase2L1CaloEGammaEmulator_cfi import phase2L1CaloEGammaEmulator

from Configuration.ProcessModifiers.premix_stage2_cff import premix_stage2
premix_stage2.toModify(phase2L1CaloEGammaEmulator,
    ecalTPEB = "DMEcalEBTriggerPrimitiveDigis",
    hcalTP = "DMHcalTriggerPrimitiveDigis",
)
```

This should maintain the era modifications for `phase2L1CaloEGammaEmulator`, but this no longer seems like a `_cfi` per se. Should this be relegated to some customization, or other python file? What's the preference for cleaning this up? Looking at the `phase2_premix` era, most files seem to have `.toModify()` to change it in some `_cfi`.

perrotta reviewed (commented):
Mostly suggestions to clean up the code, but at least the duplicate configuration files must be fixed

skkwan commented on DataFormats/L1TCalorimeterPhase2/src/classes_def.xml:
```diff
@@ -25,5 +25,22 @@
+  <class name="l1tp2::DigitizedTowerCorrelatorCollection" />
+  <class name="edm::Wrapper<l1tp2::DigitizedTowerCorrelatorCollection>" />
+
+  <class name="l1tp2::DigitizedClusterGT"  ClassVersion="4">
```
Yes so, originally I had it as ClassVersion 3, then after I made some changes to the DataFormat, there was a message that recommended changing it to ClassVersion 4.

perrotta replied:
> Yes so, originally I had it as ClassVersion 3, then after I made some changes to the DataFormat, there was a message that recommended changing it to ClassVersion 4.

In the release there is not yet such version 3, therefore please start from it

skkwan committed:
address formatting, class version, logic comments

cmsbuild commented:
Pull request #41224 was updated. @epalencia, @AdrianoDee, @srimanob, @cmsbuild, @aloeliger can you please check and sign again.

aloeliger commented:
> @aloeliger why do you want to era-modify the configuration for the base `phase2L1CaloEGammaEmulator` module, instead of just `l1tPhase2L1CaloEGammaEmulator` as it is (correctly, in my opinion) done now, and as it was done in this PR before the last commit?
> 
> What you are using is just `l1tPhase2L1CaloEGammaEmulator`, while the other one is only used as a reference, and for the validation of the configuration.
> 
> Please correct me if there is anything I miss.

@perrotta Perhaps it is my mistake, I was under the impression that you had wanted something different done than what was available right now. If the `_cfi` is acceptable as is, then I think everything is fine.

perrotta commented:
> @perrotta Perhaps it is my mistake, I was under the impression that you had wanted something different done than what was available right now. If the `_cfi` is acceptable as is, then I think everything is fine.

The problem that I saw was that there were two files identically named l1tPhase2L1CaloEGammaEmulator_cfi.py, and searchable with the same python import. That would have generated possible troubles, especially when you modify one of them and not the other one.

Now you have l1tPhase2L1CaloEGammaEmulator_cfi.py and phase2L1CaloEGammaEmulator_cfi.py, where the first one is used in the workflows and the the second one is only there as a reference and for the configuration validation: much safer!

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-81ce5a/33183/summary.html
**COMMIT**: ea0347536dfa774147e371c9f3fd4bc7f778e50c
**CMSSW**: CMSSW_13_2_X_2023-06-15-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41224/33183/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 1 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3196062
 - DQMHistoTests: Total failures: 75
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3195965
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-15-1100+81ce5a/57515/triggerResults): found differences in 3 / 46 workflows

aloeliger commented:
+l1

- @cms-sw/upgrade-l2, do you have any other comments for @skkwan?

aloeliger commented:
@AdrianoDee Are there any final comments for @skkwan? The L1T group is anxious to have this merged so that we can move on to the integration of the GT emulator which is dependent on this PR.

AdrianoDee commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
